pecos/core/ann/quantizer_impl/common.hpp

```diff
@@ -0,0 +1,237 @@
+            // zero initialization for later do_axpy
+            memset(centroids, 0, ksub * dsub * sizeof(*centroids));
+            std::vector<float> centroids_size(ksub);
+            #pragma omp parallel num_threads(threads)
+            {
+                // each thread takes care of [c_l, c_r)
+                int rank = omp_get_thread_num();
+                size_t c_l = (ksub * rank) / threads;
+                size_t c_r = (ksub * (rank + 1)) / threads;
+                for (size_t i = 0; i < X.rows; i++) {
+                    auto ci = assign[i];
+                    if (ci >= c_l && ci < c_r) {
+                        float* y = centroids + ci * dsub;
+                        const auto& xi = X.get_row(i);
+                        pecos::do_axpy(1.0, xi.val, y, dsub);
+                        centroids_size[ci] += 1;
+                    }
+                }
+                // normalize center vector
+                for (size_t ci = c_l; ci < c_r; ci++) {
+                    float* y = centroids + ci * dsub;
+                    pecos::do_scale(1.0 / centroids_size[ci], y, dsub);
+                }
+            }
+        }
+
+        inline void train(const pecos::drm_t& X_trn, int M, size_t sub_sample_points=0, int seed=0, size_t max_iter=10, int threads=32) {
+            size_t dimension = X_trn.cols;
+            if (dimension % M != 0) {
+                throw std::runtime_error("Original dimension must be divided by subspace dimension");
+            }
+            num_local_codebooks = M;
+            local_dimension = dimension / num_local_codebooks;
+            index_type n_data = X_trn.rows;
+            if (sub_sample_points == 0) {
+                sub_sample_points = n_data;
+            }
+
+            std::vector<float> centroids;
+            original_local_codebooks.resize(num_local_codebooks * num_of_local_centroids * dimension);
+            global_centroid.resize(dimension, 0);
+
+            std::vector<float> xslice(sub_sample_points * local_dimension);
+            for (index_type m = 0; m < num_local_codebooks; m++) {
+                std::vector<size_t> indices(n_data, 0);
+                std::iota(indices.data(), indices.data() + n_data, 0);
+                std::random_shuffle(indices.data(), indices.data() + n_data);
+                for (size_t i = 0; i < sub_sample_points; i++) {
+                    size_t index = indices[i];
+                    std::memcpy(xslice.data() + i * local_dimension, X_trn.val + index * dimension + m * local_dimension, local_dimension * sizeof(float));
+                }
+                pecos::drm_t Xsub;
+                Xsub.rows = sub_sample_points;
+                Xsub.cols = local_dimension;
+                Xsub.val = xslice.data();
+
+                // fit HLT or flat-Kmeans for each sub-space
+                std::vector<index_type> assignments(sub_sample_points);
+                pecos::clustering::Tree hlt(4);
```

Comment:
avoid hard-coded, try `log2(num_of_local_centroids)`.

Reply:
Resolved in next commit.